block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala

```diff
@@ -0,0 +1,254 @@
+package coop.rchain.blockstorage.dag
+
+import cats.effect.concurrent.Semaphore
+import cats.effect.{Concurrent, Sync}
+import cats.syntax.all._
+import com.google.protobuf.ByteString
+import coop.rchain.blockstorage._
+import coop.rchain.blockstorage.dag.BlockDagStorage.DeployId
+import coop.rchain.blockstorage.dag.BlockMetadataStore.BlockMetadataStore
+import coop.rchain.blockstorage.dag.EquivocationTrackerStore.EquivocationTrackerStore
+import coop.rchain.blockstorage.dag.codecs._
+import coop.rchain.blockstorage.util.BlockMessageUtil._
+import coop.rchain.casper.protocol.BlockMessage
+import coop.rchain.crypto.codec.Base16
+import coop.rchain.metrics.Metrics.Source
+import coop.rchain.metrics.{Metrics, MetricsSemaphore}
+import coop.rchain.models.BlockHash.BlockHash
+import coop.rchain.models.EquivocationRecord.SequenceNumber
+import coop.rchain.models.Validator.Validator
+import coop.rchain.models.{BlockHash, BlockMetadata, EquivocationRecord, Validator}
+import coop.rchain.shared.syntax._
+import coop.rchain.shared.{Log, LogSource}
+import coop.rchain.store.{KeyValueStoreManager, KeyValueTypedStore}
+import scodec.Codec
+import scodec.bits.ByteVector
+
+import scala.collection.immutable.SortedMap
+
+final class BlockDagKeyValueStorage[F[_]: Concurrent: Log] private (
+    lock: Semaphore[F],
+    latestMessagesIndex: KeyValueTypedStore[F, Validator, BlockHash],
+    blockMetadataIndex: BlockMetadataStore[F],
+    deployIndex: KeyValueTypedStore[F, DeployId, BlockHash],
```

Comment:
I wonder what we should do with the deploy in multiple blocks.(When a block get orphaned, the deploy would propose in another block because the block is not finalized).In the pervious implementation, the record should be there but overwritten while reading the log file.

Reply:
> In the pervious implementation, the record should be there but overwritten while reading the log file.

@zsluedem when the record is overwritten from the log? Do you mean when node is started?

I don't think we need separate persistent storage for deploy signatures at all. We can refactor this and hold latest 50 deploy signatures just in memory. When node is starting signatures can be loaded from blocks.

Reply:
When the node load the log files,  the node would load the files in sequences to write to a map in memory. So if a deploy exists in two blocks, the last index would be loaded in the map(The first one would be overwritten).

Reply:
@zsluedem this means behavior will not change. Last deploy added to deploy index will overwrite previous record.

Maybe we need to save multiple blocks with some deploy to cover all possible cases?

Reply:
I agree with covering all possible cases. It can be misleaded if the user found the deploy sometimes in one block and sometimes it is in another one.

Reply:
I opened a new bug/enhancement issue to fix this #2954.